msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AbstractApplicationBase.java

```diff
@@ -0,0 +1,476 @@
+
+package com.microsoft.aad.msal4j;
+
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.experimental.Accessors;
+import org.slf4j.Logger;
+
+import javax.net.ssl.SSLSocketFactory;
+import java.net.MalformedURLException;
+import java.net.Proxy;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Consumer;
+
+import static com.microsoft.aad.msal4j.ParameterValidationUtils.validateNotBlank;
+import static com.microsoft.aad.msal4j.ParameterValidationUtils.validateNotNull;
+
+/**
+ * Abstract class containing common methods and properties for {@link PublicClientApplication},
+ * {@link ConfidentialClientApplication}, and {@link ManagedIdentityApplication}
+ */
+public abstract class AbstractApplicationBase implements IApplicationBase {
+
+    protected Logger log;
+    protected Authority authenticationAuthority;
+
+    @Accessors(fluent = true)
+    @Getter(AccessLevel.PACKAGE)
+    protected TokenCache tokenCache;
+
+    @Accessors(fluent = true)
+    @Getter
+    private String correlationId;
+
+    @Accessors(fluent = true)
+    @Getter
+    private boolean logPii;
+
+    @Accessors(fluent = true)
+    @Getter
+    private Proxy proxy;
+
+    @Accessors(fluent = true)
+    @Getter
+    private SSLSocketFactory sslSocketFactory;
+
+    @Accessors(fluent = true)
+    @Getter(AccessLevel.PACKAGE)
+    private ServiceBundle serviceBundle;
+
+    @Accessors(fluent = true)
+    @Getter(AccessLevel.PACKAGE)
+    private Consumer<List<HashMap<String, String>>> telemetryConsumer;
+
+    @Accessors(fluent = true)
+    @Getter
+    private IHttpClient httpClient;
+
+    @Accessors(fluent = true)
+    @Getter
+    private Integer connectTimeoutForDefaultHttpClient;
+
+    @Accessors(fluent = true)
+    @Getter
+    private Integer readTimeoutForDefaultHttpClient;
+
+    @Accessors(fluent = true)
+    @Getter
+    private String applicationName;
+
+    @Accessors(fluent = true)
+    @Getter
+    private String applicationVersion;
+
+    //The following fields are set in only some applications and/or set internally by the library. To avoid excessive
+    // type casting throughout the library they are defined here as package-private, but will not be part of this class's Builder
+    @Accessors(fluent = true)
+    @Getter(AccessLevel.PACKAGE)
+    private boolean validateAuthority;
+
+    @Accessors(fluent = true)
+    @Getter(AccessLevel.PACKAGE)
+    private String clientId;
+
+    @Accessors(fluent = true)
+    @Getter(AccessLevel.PACKAGE)
+    private String authority;
+
+    @Accessors(fluent = true)
+    @Getter(AccessLevel.PACKAGE)
+    private boolean autoDetectRegion;
+
+    @Accessors(fluent = true)
+    @Getter(AccessLevel.PACKAGE)
+    protected String azureRegion;
+
+    @Accessors(fluent = true)
+    @Getter(AccessLevel.PACKAGE)
+    private boolean instanceDiscovery;
```

Comment:
I don't think we need `instanceDiscovery`, `autoDetection` and `azureRegion` in the base.

Reply:
I originally put these here to avoid excessive casting in places where the Application class was referenced via an MsalRequest object. However, since this casting would only occur in the AadInstanceDiscoveryProvider class, and it'd be confusing to have these field in AbstractApplicationBase but never used, in the latest commit I moved them back to AbstractClientApplicationBase and just did a bunch of type casting in AadInstanceDiscoveryProvider.